### PR TITLE
Java unwind enum static

### PR DESCRIPTION
**Pull request by mgudemann**
unwind loops in `<clinit>` for `#enum` times as heuristic to make static initialization of Java enums finish

requires #425 

[real diff](https://github.com/diffblue/cbmc/pull/460/commits/ed80984839d29de7e5dcfeeb7c4cf45b5bf70e04)

### CODE REVIEW

File: src/cbmc/cbmc_parse_options.cpp [hunk 1]
```diff
@@ -430,6 +434,81 @@ void cbmc_parse_optionst::get_command_line_options(optionst &options)
...
+  if(vla_length!="")
+    assert(
+      max_vla_length>=unwind_max &&
+      "cannot unwind <clinit> due to insufficient max vla length");
+}
+
+/*******************************************************************\
+
+/*******************************************************************\
+
```
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp]: Duplicate comment-start-markers

File: src/cbmc/cbmc_parse_options.cpp [hunk 2]
```diff
@@ -430,6 +434,81 @@ void cbmc_parse_optionst::get_command_line_options(optionst &options)
...
+              id2string(it.function)+"."+std::to_string(loop_id)+":"+
+              std::to_string(unwinds);
+            options.set_option("unwindset", set+newset);
+          }
+        }
+      }
+    }
+  }
+  const std::string &vla_length=options.get_option("java-max-vla-length");
+  size_t max_vla_length=vla_length==""?0:unsafe_string2unsigned(vla_length);
```
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp]: As you are using `unsafe_...` you don't actually need to have a `vla_length==""?0` guard as `0` will be the default return value.

File: src/cbmc/cbmc_parse_options.cpp [hunk 3]
```diff
@@ -430,6 +434,81 @@ void cbmc_parse_optionst::get_command_line_options(optionst &options)
...
+  if(vla_length!="")
+    assert(
+      max_vla_length>=unwind_max &&
+      "cannot unwind <clinit> due to insufficient max vla length");
```
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp]: In all this you might actually play a bit safer:
```
const std::string &vla_length=options.get_option("java-max-vla-length");
if(!vla_length.empty())
{
  size_t max_vla_length=safe_string2unsigned(vla_length);
  if(max_vla_length<unwind_max)
    throw "cannot unwind <clinit> due to insufficient max vla length";
}
```

File: src/cbmc/cbmc_parse_options.cpp [hunk 4]
```diff
@@ -430,6 +434,81 @@ void cbmc_parse_optionst::get_command_line_options(optionst &options)
...
+
+void cbmc_parse_optionst::unwind_enum_static(
+  const goto_functionst &goto_functions,
+  optionst &options)
+{
+  size_t unwind_max=0;
+  forall_goto_functions(f, goto_functions)
+  {
+    auto &p=f->second.body;
+    for(auto it : p.instructions)
```
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp]: Shouldn't that be `const auto &ins` (i.e, const-ness, reference, and it isn't an iterator hence "it" is not the best choice of a name); on a second thought, const-ness might actually be implied in "auto" here.

Comment (mgudemann) [inline src/cbmc/cbmc_parse_options.cpp]: ok, is `const auto &ins` now

File: src/cbmc/cbmc_parse_options.cpp (hunk 2 above)
Comment (mgudemann) [inline src/cbmc/cbmc_parse_options.cpp]: ok, removed

File: src/cbmc/cbmc_parse_options.cpp (hunk 3 above)
Comment (mgudemann) [inline src/cbmc/cbmc_parse_options.cpp]: changed to your suggestion

File: src/cbmc/cbmc_parse_options.cpp (hunk 1 above)
Comment (mgudemann) [inline src/cbmc/cbmc_parse_options.cpp]: done

**Comment by tautschnig** [comment]
Thanks a lot for all the updates. I am going to look at #425 first, and once that's merged (and this one rebased) I'll take another look here if that's ok?

**Comment by mgudemann** [comment]
@tautschnig of course, your comments are appreciated

File: src/cbmc/cbmc_parse_options.cpp [hunk 5]
```diff
@@ -430,6 +434,80 @@ void cbmc_parse_optionst::get_command_line_options(optionst &options)
 
 /*******************************************************************\
 
+Function: cbmc_parse_optionst::unwind_enum_static
```
Comment (peterschrammel) [inline src/cbmc/cbmc_parse_options.cpp]: That's not the ideal place for this code. 
Can you move this into goto-programs/ in the style of the remove_* passes there?

Comment (mgudemann) [inline src/cbmc/cbmc_parse_options.cpp]: there's `goto-program/remove_static_init_loop.{cpp,h}` now

**Comment by mgudemann** [comment]
and rebased...
linter complains about two lines that are too long, there are the `help()` output with description for the CLI parameters

**Comment by tautschnig** [comment]
@mgudemann Please add `// NOLINT(whitespace/line_length)` at the end of those lines or `// NOLINTNEXTLINE(whitespace/line_length)` in the line before those being complained about.

**Comment by mgudemann** [comment]
added `// NOLINTNEXTLINE`

File: src/goto-programs/remove_static_init_loops.cpp [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  const std::string &vla_length=options.get_option("java-max-vla-length");
+  if(!vla_length.empty())
+  {
+    size_t max_vla_length=safe_string2unsigned(vla_length);
+    if(max_vla_length<unwind_max)
+      throw "cannot unwind <clinit> due to insufficient max vla length";
+  }
+}
+
+void remove_static_init_loops(
```
Comment (tautschnig) [inline src/goto-programs/remove_static_init_loops.cpp]: Would you mind adding a function comment header here (I'm surprised the linter doesn't complain)?

Comment (mgudemann) [inline src/goto-programs/remove_static_init_loops.cpp]: done

**Pull request merged**